agent/taskresource/asmauth/asmauth_test.go

```diff
@@ -0,0 +1,143 @@
+	taskARN                = "task1"
+)
+
+var (
+	asmAuthDataVal       string
+	requiredASMResources []*apicontainer.ASMAuthData
+)
+
+func init() {
+	asmAuthDataBytes, _ := json.Marshal(&asm.AuthDataValue{
+		Username: aws.String(username),
+		Password: aws.String(password),
+	})
+	asmAuthDataVal = string(asmAuthDataBytes)
+	requiredASMResources = []*apicontainer.ASMAuthData{
+		{
+			CredentialsParameter: secretID,
+			Region:               region,
+		},
+	}
+}
+
+func TestCreateAndGet(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	credentialsManager := mock_credentials.NewMockManager(ctrl)
+	asmClientCreator := mock_factory.NewMockClientCreator(ctrl)
+	mockASMClient := mock_secretsmanageriface.NewMockSecretsManagerAPI(ctrl)
+
+	iamRoleCreds := credentials.IAMRoleCredentials{}
+	creds := credentials.TaskIAMRoleCredentials{
+		IAMRoleCredentials: iamRoleCreds,
+	}
+	asmSecretValue := &secretsmanager.GetSecretValueOutput{
+		SecretString: aws.String(asmAuthDataVal),
+	}
+	gomock.InOrder(
+		credentialsManager.EXPECT().GetTaskCredentials(executionCredentialsID).Return(creds, true),
+		asmClientCreator.EXPECT().NewASMClient(region, iamRoleCreds).Return(mockASMClient),
+		mockASMClient.EXPECT().GetSecretValue(gomock.Any()).Do(func(in *secretsmanager.GetSecretValueInput) {
+			assert.Equal(t, aws.StringValue(in.SecretId), secretID)
+		}).Return(asmSecretValue, nil),
+	)
+	asmRes := &ASMAuthResource{
+		executionCredentialsID: executionCredentialsID,
+		requiredASMResources:   requiredASMResources,
+		credentialsManager:     credentialsManager,
+		asmClientCreator:       asmClientCreator,
+	}
+	require.NoError(t, asmRes.Create())
+	dac, ok := asmRes.GetASMDockerAuthConfig(secretID)
+	require.True(t, ok)
+	assert.Equal(t, dac.Username, username)
+	assert.Equal(t, dac.Password, password)
+}
+
+func TestMarshalUnmarshalJSON(t *testing.T) {
+	asmResIn := &ASMAuthResource{
+		taskARN:                taskARN,
+		executionCredentialsID: executionCredentialsID,
+		createdAt:              time.Now(),
+		knownStatusUnsafe:      taskresource.ResourceCreated,
+		desiredStatusUnsafe:    taskresource.ResourceCreated,
+		requiredASMResources:   requiredASMResources,
+	}
+
+	bytes, err := json.Marshal(asmResIn)
+	require.NoError(t, err)
+
+	asmResOut := &ASMAuthResource{}
+	err = json.Unmarshal(bytes, asmResOut)
+	require.NoError(t, err)
+	assert.Equal(t, asmResIn.taskARN, asmResOut.taskARN)
+	assert.WithinDuration(t, asmResIn.createdAt, asmResOut.createdAt, time.Microsecond)
+	assert.Equal(t, asmResIn.desiredStatusUnsafe, asmResOut.desiredStatusUnsafe)
+	assert.Equal(t, asmResIn.knownStatusUnsafe, asmResOut.knownStatusUnsafe)
+	assert.Equal(t, asmResIn.executionCredentialsID, asmResOut.executionCredentialsID)
+	assert.Equal(t, len(asmResIn.requiredASMResources), len(asmResOut.requiredASMResources))
+	assert.Equal(t, asmResIn.requiredASMResources[0], asmResOut.requiredASMResources[0])
+}
+
+func TestInitialize(t *testing.T) {
```

Comment:
Do we test what happens when we initialize twice? Might be more appropriate for integ test

Reply:
🤔

Reply:
IE: what happens when we initialize, agent gets shut down, and then agent starts back up?

Reply:
> what happens when we initialize, agent gets shut down, and then agent starts back up

so initialize is part of resource provisioning, and if the agent is restarted the resources will be reinitialized as part of the task engine's `synchronizeState()` workflow.